lib/fsm.ex

```diff
@@ -13,10 +13,18 @@ defmodule Fsm do
         %__MODULE__{}
       end
 
-      def new(initial_state) do
+      def new(state: initial_state) do
         %{%__MODULE__{} | state: initial_state}
       end
 
+      def new(data: initial_data) do
+        %{%__MODULE__{} | data: initial_data}
+      end
+
+      def new(state: initial_state, data: initial_data) do
+        %{%__MODULE__{} | state: initial_state, data: initial_data}
+      end
+
```

Comment:
I think you could replace all four definitions of `new` with the single one:

```elixir
def new(params \\ []), do:
  struct!(__MODULE__, params)
```

Could you please give it a try?